### DISCUSSION TITLE
Different Markdown options based on content language?

### DISCUSSION THREAD

**Discussion by saadatm** [11ty/eleventy#2891] [Q&A] [answered]
In a multilingual Eleventy website, it it possible to specify different Markdown options for each (or a particular) content language?

Here’s a use case: Suppose that we have a website in two languages — English and Urdu. Both languages contain content written in Markdown. Because Urdu is an RTL language, its smart/curly quotes are opposite of what are used in English:

* In English, the opening double quote is `“` and the closing double quote is `”`
* In Urdu, the opening double quote is `”` and the closing double quote is `“`

`markdown-it` lets us specify the double (and single) quotes replacement pairs via its `quotes` option. But doing so in the Eleventy configuration file will use those pairs for English and Urdu _both_. Can we somehow specify a different value for `quotes` (or any other Markdown option) for only Urdu?

**Comment by pdehaan** [comment]
@saadatm I don't think you can have multiple Markdown engines, but you could create 1+ shortcodes which parses a Markdown block with the smart quotes of your choice:

```js
// eleventy.config.js (Eleventy v2)
const markdownIt = require("markdown-it");

/**
 * @param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig
 * @returns {ReturnType<import("@11ty/eleventy/src/defaultConfig")>}
 */
module.exports = function (eleventyConfig) {
  // general/shared markdown-it config options.
  const mdOptions = {
    html: true,
    breaks: true,
    linkify: true,
    typographer: true,
  };
  // override quotes per language/locale.
  const mdEngines = {
    english: markdownIt({ ...mdOptions, quotes: '“”‘’' }),
    german: markdownIt({ ...mdOptions, quotes: '„“‚‘' }),
    urdu: markdownIt({ ...mdOptions, quotes: '”“’‘' }),
  };

  // set default markdown library to use English quotes/config.
  eleventyConfig.setLibrary("md", mdEngines.english);
  // paired shortcode to convert markdown to HTML using the specified markdown engine language.
  eleventyConfig.addPairedShortcode("markdown", function (content, language="english") {
    // throw hard Error on unknown language.
    if (!(language in mdEngines)) {
      throw new Error(`Unknown markdown engine language: "${language}". Expected one of: ${JSON.stringify(Object.keys(mdEngines))}`);
    }
    const engine = mdEngines[language];
    const html = engine.render(content).trim();
    return `
      <div data-language="${language}">${html}</div>
    `.trim();
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

```liquid
---
# src/index.liquid
title: This is a "fancy" title!
author: Brian O'Malley
---

{% markdown "german" -%}
  # {{ title }}
  {{ author }}
{% endmarkdown -%}

{% markdown "urdu" -%}
  # {{ title }}
  {{ author }}
{% endmarkdown -%}
```

### OUTPUT
```html
<div data-language="german">
  <h1>This is a „fancy“ title!</h1>
  <p>Brian O’Malley</p>
</div>

<div data-language="urdu">
  <h1>This is a ”fancy“ title!</h1>
  <p>Brian O’Malley</p>
</div>
```

I imagine you could change the language names to codes/locales if that works better for you, with the general caveats of it might get slightly trickier if you're using hyphens in names instead of underscores since "en-US" style codes aren't valid JavaScript variable names, so you'd have to use `['en-US']` in a few places. Although maybe not relevant if your locale is "ur". :shrug:

**Reply by pdehaan** [reply]
And if you use an invalid "language" for the shortcode (for example, "ur" instead of "urdu"), you should see a fatal error like this buried in your terminal:

```text
Unknown markdown engine language: "ur". Expected one of: ["english","german","urdu"], file:./src/index.liquid, line:7, col:1 (via RenderError)
```

You could change that to use warnings and default to English if you prefer, but I like fatal errors so I know one of my configs somewhere is wrong and not silently doing the wrong thing.

**Reply by saadatm** [reply]
Hmm. I guess I'll always need to wrap Urdu posts/pages in such a shortcode, then.

Thanks for the code and your help, @pdehaan. 🙏

**Reply by darthmall** [reply]
Alternatively you could use the `<q></q>` element: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/q

This should allow you to set the desired typographic symbols using CSS (if the browser doesn’t do so for you automatically based on your `lang` attribute).

**Reply by pdehaan** [reply]
Great idea @darthmall! I did a quick test locally and it seems to work perfectly (if you set your `html[lang]`):
```html
<html lang="de"><head></head><body>

<main>Here is some <q>Quoted?</q> text. So fancy!</main>
    
</body></html>
```

**Reply by darthmall** [reply]
One thing I should've mentioned is that this element is intended only for quotations of other sources. So it wouldn't be appropriate if you were doing something like defining a term—e.g. “the definition of ‘sesquipedalian’ is…”

**Reply by saadatm** [reply]
@darthmall Aha, yes! This will work beautifully for quotes.

However, having multiple `markdown-it` instances (or shortcodes, as suggested by @pdehaan) will help beyond the selection of correct quotes. For example, in the `markdown-it-footnote` extension, we can [replace its template functions](https://github.com/markdown-it/markdown-it-footnote#customize), and a separate `markdown-it` instance or shortcode will let us use localized strings for footnotes in Urdu posts.

**Comment by rothsandro** [comment, marked as answer]
Maybe you could create a wrapper around markdown-it and provide your wrapper to Eleventy? 

```js
const markdownIt = require("markdown-it");

module.exports = function (eleventyConfig) {
  const libEnglish = markdownIt({ html: true });
  const libUrdu = markdownIt({ html: false });

  eleventyConfig.setLibrary("md", {
    ...libEnglish,
    render: (content, ctx) => {
      const lib = ctx.lang === "urdu" ? libUrdu : libEnglish;
      return lib.render(content, ctx);
    },
  });
};
```

The `ctx` variable gives you access to front matter data and other props like inputPath that you can use to determine the language. Beside the `render` property there are one or two other props that Eleventy uses (as far as I remember) which you may need to overwrite but should be possible.

**Reply by pdehaan** [reply]
Great solution, @rothsandro!
That worked great locally for me in a test repo with some directory data files where I could try setting the `lang` property for groups of files.

**Reply by saadatm** [reply]
Wonderful! This is working great, @rothsandro! Thank you so much. 🙏
